### ISSUE TITLE
Try to replace our .envrc approach with lorri

### ISSUE THREAD

**Issue by mpscholten** [digitallyinduced/ihp#339]

**Comment by ruhatch** [comment]
I've looked into this a little bit and it seems fruitful. The main issue is the need to have the `lorri watch` command running, which adds some overhead for users.

It's easy on NixOS, because there's an option `services.lorri.enable` to set up the `lorri` daemon. Any thoughts about how to do this on other platforms though? We can do the same install steps as for `direnv` and `cachix`, but I'm not sure about running it.

**Comment by ruhatch** [comment]
> Maybe we could try to add the `lorri watch` to the `./start` script in the project dir (except for nixos where we could start it only if it's not already started)?

Yeah, that sounds like a good plan. I wonder whether we could then use that to automate the 'stop server, make .envrc, start server' stuff. Perhaps the script notices when a later evaluation becomes available and automatically restarts or something.

**Comment by ruhatch** [comment]
Steps:
- [ ] Update `ihp-new` to install `lorri` and run `lorri init`
- [ ] Update project boilerplate to include `shell.nix` for use by `lorri`
- [ ] Include `lorri watch` in `./start` script (don't run if `lorri daemon` process is detected)
- [ ] Update documentation for new projects and package management, plus extra suggestion for NixOS users to use `configuration.nix`

**Comment by mpscholten** [comment]
Thanks for tackling this :)

Auto-restarting the server has been in the backlog of IHP actually. But it didn't make it in time for the open source release, so we dropped it. It would be great to add this, because then we can remove a few steps in the documentation, which makes it a lot easier to work with IHP.

On thing we'll have to deal with: When `lorri` is updating the nix evaluation and the dev server is running, it will try to access the unix socket of the dev server. [This is a common problem.](https://github.com/digitallyinduced/ihp/wiki/Troubleshooting#error-file-builddbspgsql5432-has-an-unsupported-type). We'd need to change nix so, that it will not load this unix socket.

We also need to think about how we can update existing projects (Installing `lorri`, intializing it, removing the old .envrc to avoid issues, updating the `start` script).

**Comment by ruhatch** [comment]
> Thanks for tackling this :)
> 
> Auto-restarting the server has been in the backlog of IHP actually. But it didn't make it in time for the open source release, so we dropped it. It would be great to add this, because then we can remove a few steps in the documentation, which makes it a lot easier to work with IHP.
> 
> On thing we'll have to deal with: When `lorri` is updating the nix evaluation and the dev server is running, it will try to access the unix socket of the dev server. [This is a common problem.](https://github.com/digitallyinduced/ihp/wiki/Troubleshooting#error-file-builddbspgsql5432-has-an-unsupported-type). We'd need to change nix so, that it will not load this unix socket.

I'm confused, because it looks like you should be filtering out the `build` folder with the call to `gitignoreSource` in `NixSupport/default.nix`. Would you expect that to avoid the `build/db` problem?

> We also need to think about how we can update existing projects (Installing `lorri`, intializing it, removing the old .envrc to avoid issues, updating the `start` script).

Yep, added this to the list above, interesting issue :P

**Comment by mpscholten** [comment]
Actually it should fix the problem, but somehow it doesn’t work. You can test it out by running a terminal with ./start and after that starting a second with nix-shell. The nix-shell will fail to start.

**Comment by ruhatch** [comment]
I figured that out - basically this line in `NixSupport/default.nix` uses the directory in string interpolation:
```
    pkgs = import "${projectPath}/Config/nix/nixpkgs-config.nix" { ihp = ihp; };
```
Apparently, because the `projectPath` has a path type, this evaluates the whole source tree! Simply adding `toString` fixes this:
```
    pkgs = import "${toString projectPath}/Config/nix/nixpkgs-config.nix" { ihp = ihp; };
```
See [this](https://stackoverflow.com/questions/43850371/when-does-a-nix-path-type-make-it-into-the-nix-store-and-when-not) for explanation. One to watch out for!

**Comment by mpscholten** [comment]
Very cool :) Thanks for finding out! Just tried it locally and it indeed fixes the issue. Pushed it via https://github.com/digitallyinduced/ihp/commit/c03343276f4490f979da37c1bad4a9ec29bc8218 

Then we can actually remove https://github.com/digitallyinduced/ihp/wiki/Troubleshooting#error-file-builddbspgsql5432-has-an-unsupported-type from the wiki with the next IHP release :)
